libsql_client/dbapi2/types.py

```diff
@@ -0,0 +1,1423 @@
+        _Parameters as SqlParameters,
+        _SqliteData as SqlData,
+    )
+else:
+    SqlParameters = Any
+    SqlData = Any
+
+PathLike = Union[str, bytes, os.PathLike]
+IsolationLevel = Literal["", "DEFERRED", "EXCLUSIVE", "IMMEDIATE"]
+isolation_level_set: Set[IsolationLevel] = {
+    "", "DEFERRED", "EXCLUSIVE", "IMMEDIATE",
+}
+
+SqlNativeType = Union[None, int, float, str, bytes]
+ConnectionTypes = Union[sqlite3.dbapi2.Connection, "Connection"]
+ConnectFactory = Callable[..., ConnectionTypes]
+CursorFactory = Callable[["Connection"], "Cursor"]
+RowFactory = Callable[["Cursor", Iterable[Any]], Any]
+TextFactory = Callable[[bytes], Any]
+ConverterCallback = Callable[[Any], Any]
+
+AuthorizerCallback = Callable[
+    ...,
+    int  # SQLITE_OK, SQLITE_DENY, or SQLITE_IGNORE
+]
+ProgressHandler = Callable[[], int]
+TraceCallback = Callable[[str], None]
+BackupProgressCallback = Callable[[int, int, int], None]
+
+RawResults = proto.StmtResult
+RawColumn = proto.Col
+RawRow = List[proto.Value]
+
+LEGACY_TRANSACTION_CONTROL: Literal[-1] = -1
+Autocommit = Union[bool, Literal[-1]]
+
+
+def _get_local_default_limits() -> Dict[int, int]:
+    if sys.version_info[:2] < (3, 11):
+        return {}
+    # while this doesn't make much sense, let's keep it compatible and
+    # allows tests to pass, such as checking if SQLITE_LIMIT_SQL_LENGTH
+    # is being enforced
+    with sqlite3.dbapi2.Connection(":memory:") as con:
+        categories = (
+            "SQLITE_LIMIT_LENGTH",
+            "SQLITE_LIMIT_SQL_LENGTH",
+            "SQLITE_LIMIT_COLUMN",
+            "SQLITE_LIMIT_EXPR_DEPTH",
+            "SQLITE_LIMIT_COMPOUND_SELECT",
+            "SQLITE_LIMIT_VDBE_OP",
+            "SQLITE_LIMIT_FUNCTION_ARG",
+            "SQLITE_LIMIT_ATTACHED",
+            "SQLITE_LIMIT_LIKE_PATTERN_LENGTH",
+            "SQLITE_LIMIT_VARIABLE_NUMBER",
+            "SQLITE_LIMIT_TRIGGER_DEPTH",
+            "SQLITE_LIMIT_WORKER_THREADS",
+        )
+        defaults = {}
+        for category in categories:
+            cat_id = getattr(_reexports, category)
+            defaults[cat_id] = con.getlimit(cat_id)
+        return defaults
+
+
+_default_limits = _get_local_default_limits()
+
+
+def check_valid_autocommit(value: Any) -> Autocommit:
+    ":meta private:"
+    if value is LEGACY_TRANSACTION_CONTROL or isinstance(value, bool):
+        return value
+    raise ValueError("autocommit must be True, False, or "
+                     "sqlite3.LEGACY_TRANSACTION_CONTROL")
+
+
+_callback_tracebacks: bool = False
+
+
+def enable_callback_tracebacks(flag: bool) -> None:
+    """See :py:func:`sqlite3.enable_callback_tracebacks`"""
+    global _callback_tracebacks
+    _callback_tracebacks = flag
+
+
+class RawExecuteResult:
+    """Iterable over pairs of ``result, error``.
+
+    :meta private:
+    """
+    results: List[Optional[RawResults]]
+    errors: List[Optional[BaseException]]
+    _idx: int
+    __slots__ = ("results", "errors", "_idx")
```

Comment:
because you return two lists in your API, this class makes it easy to use that without bothering.

Although we have a `hrana.py` and `types.py`, since all will be hrana, it's nore like `websockets.py` and `types.py` composed of hrana stuff. But I won't rename these until we come to implement the http... if websockets is gone, then I'll just replace the contents of hrana.py, if we keep both, I'll likely rename the file.